### PR TITLE
Add support for platform libraries

### PR DESCRIPTION
**Pull request by ezyang**
Subsumes #2540.

CC @edsko and @dcoutts

### CODE REVIEW

**Comment by ezyang** [comment]
cabal-install should be working now.

File: Cabal/doc/developing-packages.rst [hunk 1]
```diff
@@ -1432,6 +1432,94 @@ runner:
...
+      hs_exit();
+    }
+
+With modern ghc regular libraries are installed in directories that contain
+package keys. This isn't usually a problem because the package gets registered
+in ghc's package DB and so we can figure out what the location of the library
+is. Foreign libraries however don't get registered, which means that we'd have
+to way of finding out where a platform library got installed (other than by
+searching the ``lib/`` directory). Instead, we install foreign libraries in
+``~/.cabal/lib``, must like we install executables in ``~/.cabal/bin``.
```
Comment (abooij) [inline Cabal/doc/developing-packages.rst]: must -> much

**Comment by ezyang** [comment]
Failure is because I need to find the gcc executable correctly on Windows.

**Comment by ezyang** [comment]
If there aren't any objections, I will merge.

**Comment by 23Skidoo** [comment]
Can we add a changelog note?

**Comment by ezyang** [comment]
Added, and merged.

**Pull request closed without merging**